File: LicenseManager_VB/frmMain.vb
Imports System.Data 'DataTable
Imports System.Reflection 'Assembly
Imports System.IO 'Directory
Imports System.Collections.Generic 'Dictionary
Imports Utility 'ILicense & IData

Public Class frmMain

    Private da As Assembly = Nothing

    Private Sub frmMain_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        Dim dt As New DataTable()

        dt.Columns.Add("Key")
        dt.Columns.Add("Value")

        dgvOptions.DataSource = dt

        Try
            Dim assembly As Assembly = assembly.GetExecutingAssembly()
            Dim manifestModule As [Module] = assembly.ManifestModule
            Dim peKind As PortableExecutableKinds
            Dim machine As ImageFileMachine
            manifestModule.GetPEKind(peKind, machine)

            'load x32 or x64 managed wrapper which depends on directory
            If machine.ToString.Contains("86") Then
                da = assembly.LoadFrom(Directory.GetCurrentDirectory() + "\Utility_x86.dll") 'looking for x86 version
            Else
                da = assembly.LoadFrom(Directory.GetCurrentDirectory() + "\Utility_x64.dll") 'looking for x64 version
            End If
        Catch ex As Exception
            System.Diagnostics.Trace.Write(ex.Message)
            btnEncode.Enabled = False
            btnDecode.Enabled = False
        End Try
    End Sub

    Private Sub btnEncode_Click(sender As System.Object, e As System.EventArgs) Handles btnEncode.Click
        If da IsNot Nothing Then
            Dim lt As Type = da.GetType("Utility.License") 'get License type from loaded assembly
            Dim ld As ILicense = DirectCast(Activator.CreateInstance(lt), ILicense) 'instantiate interface

            ld.Customer = tbCustomer.Text
            ld.Product = tbProduct.Text
            ld.Version = tbVersion.Text

            Dim dt As DataTable = DirectCast(dgvOptions.DataSource, DataTable)
            'ld.D = New IData(dt.Rows.Count - 1) {}
            ld.Dict = New Dictionary(Of String, String)()

            For i As Integer = 0 To dt.Rows.Count - 1
                'Dim dataType As Type = da.GetType("Utility.Data") 'get License type from loaded assembly
                'Dim data As IData = DirectCast(Activator.CreateInstance(dataType), IData) 'instantiate interface

                'data.Key = dt.Rows(i)("Key").ToString
                'data.Value = dt.Rows(i)("Value").ToString

                'ld.D(i) = data
                Try
                    ld.Dict.Add(dt.Rows(i)("Key").ToString, dt.Rows(i)("Value").ToString)
                Catch ex As Exception
                    MessageBox.Show("Key Column MUST be unique", "Error", MessageBoxButtons.OK)
                End Try
            Next

            tbLicense.Text = ld.Save()
        End If
    End Sub

    Private Sub btnDecode_Click(sender As System.Object, e As System.EventArgs) Handles btnDecode.Click
        If da IsNot Nothing Then
            Dim lt As Type = da.GetType("Utility.License") 'get License type from loaded assembly
            Dim ld As ILicense = DirectCast(Activator.CreateInstance(lt), ILicense) 'instantiate interface

            Try
                ld.Open(tbLicense.Text)
            Catch ex As Exception
                MessageBox.Show(ex.Message, "Error")
            End Try

            tbCustomer.Text = ld.Customer
            tbProduct.Text = ld.Product
            tbVersion.Text = ld.Version

            Dim dt As New DataTable
            dt.Columns.Add("Key")
            dt.Columns.Add("Value")

            'For Each l As IData In ld.D
            '    dt.Rows.Add(l.Key, l.Value)
            'Next

            For Each item As KeyValuePair(Of String, String) In ld.Dict
                dt.Rows.Add(item.Key, item.Value)
            Next

            dgvOptions.DataSource = dt
        End If
    End Sub

End Class

File: LicenseManager/frmMain.cs
using System;
using System.Windows.Forms;
using System.Data; //DataTable
using System.Reflection; //Assembly
using System.IO; //Directory
using Utility; //ILicense & IData

namespace LicenseScheme
{
    public partial class frmMain : Form
    {
        private ILicense ld = null;
        private Type dataType = null;

        public frmMain()
        {
            InitializeComponent();

            //setup grid
            DataTable dt = new DataTable();
            dt.Columns.Add("Key");
            dt.Columns.Add("Value");

            dgvOptions.DataSource = dt;

            Assembly da = null;
            try
            {
                da = Assembly.LoadFrom(Directory.GetCurrentDirectory() + @"\Utility_x86.dll"); //looking for x32 version                    
            }
            catch(Exception)
            {
                da = Assembly.LoadFrom(Directory.GetCurrentDirectory() + @"\Utility_x64.dll"); //looking for x64 version
            }

            if (da != null)
            {
                Type lt = da.GetType("Utility.License"); //get License type from loaded assembly
                ld = (ILicense)Activator.CreateInstance(lt); //instantiate interface

                dataType = da.GetType("Utility.Data"); //get Data type from loaded assembly                
            }
        }
        
        private void btnEncode_Click(object sender, EventArgs e)
        {
            //check for properly loaded assembly before we try to use
            if (ld != null)
            {
                ld.Customer = tbCustomer.Text;
                ld.Product = tbProduct.Text;
                ld.Version = tbVersion.Text;
                
                DataTable dt = (DataTable)dgvOptions.DataSource;
                ld.D = new IData[dt.Rows.Count];
                
                for (int i = 0; i < dt.Rows.Count; i++)
                {
                    IData data = (IData)Activator.CreateInstance(dataType); //instantiate interface

                    data.Key = dt.Rows[i]["Key"].ToString();
                    data.Value = dt.Rows[i]["Value"].ToString();

                    ld.D[i] = data;
                }

                try
                {
                    tbLicense.Text = ld.Save();
                }
                catch (Exception ex)
                {
                    MessageBox.Show(ex.Message, "Error");
                }
            }
        }

        private void btnDecode_Click(object sender, EventArgs e)
        {
            //check for properly loaded assembly before we try to use
            if (ld != null)
            {
                try
                {
                    ld.Open(tbLicense.Text);
                }
                catch (Exception ex)
                {
                    MessageBox.Show(ex.Message, "Error");
                }

                tbCustomer.Text = ld.Customer;
                tbProduct.Text = ld.Product;
                tbVersion.Text = ld.Version;

                DataTable dt = new DataTable();
                dt.Columns.Add("Key");
                dt.Columns.Add("Value");

                foreach (IData l in ld.D)
                    dt.Rows.Add(l.Key, l.Value);

                dgvOptions.DataSource = dt;
            }
        }
    }
}
